File: backend/app/schemas/assignments.py
from __future__ import annotations

from pydantic import BaseModel, Field


class CodeFile(BaseModel):
    path: str
    content: str = ""


class AssignmentAnalysisRequest(BaseModel):
    assignment_id: str | None = None
    assignment_title: str
    course_id: str | None = None
    class_id: str | None = None
    student_id: str | None = None
    rubric_id: str | None = None
    repository_url: str | None = None
    description: str | None = None
    files: list[CodeFile] = Field(default_factory=list)


class AssignmentCreateRequest(BaseModel):
    assignment_id: str | None = None
    title: str
    course_id: str = "course_web_2026"
    class_id: str = "class_cs_2024_01"
    description: str | None = None
    rubric_id: str | None = None


class AssignmentItem(BaseModel):
    assignment_id: str
    title: str
    course_id: str
    course_name: str
    class_id: str
    class_name: str
    description: str
    rubric_id: str | None = None
    created_at: str
    submitted_count: int = 0
    access_scope: str = "demo"


class AssignmentListResponse(BaseModel):
    assignments: list[AssignmentItem]


class AssignmentExportResponse(BaseModel):
    assignment_id: str
    filename: str
    content_type: str = "text/markdown; charset=utf-8"
    markdown: str
    generated_at: str
    access_scope: str = "demo"


class AssignmentScore(BaseModel):
    dimension: str
    score: int
    summary: str
    evidence: list[str] = Field(default_factory=list)


class AssignmentFinding(BaseModel):
    severity: str
    title: str
    detail: str
    suggestion: str


class CodeEvidenceSnippet(BaseModel):
    path: str
    module: str
    capability: str
    line_start: int
    line_end: int
    snippet: str


class CapabilityEvidence(BaseModel):
    dimension: str
    evidence: str
    source: str


class AnalysisTraceStep(BaseModel):
    node: str
    title: str
    status: str
    summary: str
    evidence: list[str] = Field(default_factory=list)


class CodeStructureSummary(BaseModel):
    file_count: int
    entry_files: list[str] = Field(default_factory=list)
    test_files: list[str] = Field(default_factory=list)
    documentation_files: list[str] = Field(default_factory=list)
    config_files: list[str] = Field(default_factory=list)
    detected_frameworks: list[str] = Field(default_factory=list)
    detected_capabilities: list[str] = Field(default_factory=list)
    risk_signals: list[str] = Field(default_factory=list)


class Citation(BaseModel):
    title: str
    source_type: str
    snippet: str


class AssignmentReportSummary(BaseModel):
    report_id: str
    student_id: str
    student_name: str
    overall_score: int
    status: str
    summary: str


class AssignmentDashboardMetric(BaseModel):
    label: str
    value: str
    trend: str | None = None


class TeachingSuggestion(BaseModel):
    knowledge_point: str
    class_evidence: str
    suggested_activity: str
    practice_task: str
    expected_improvement: str


class AssignmentAnomaly(BaseModel):
    severity: str
    title: str
    affected_students: list[str] = Field(default_factory=list)
    evidence: str
    suggested_action: str


class AbilityHeatmapCell(BaseModel):
    student_id: str
    student_name: str
    dimension: str
    score: int
    level: str


class DirectionDistributionItem(BaseModel):
    direction: str
    count: int
    ratio: float


class DataCoverageMetric(BaseModel):
    label: str
    covered: int
    total: int
    ratio: float


class ClassAbilityProfile(BaseModel):
    heatmap: list[AbilityHeatmapCell] = Field(default_factory=list)
    direction_distribution: list[DirectionDistributionItem] = Field(default_factory=list)
    data_coverage: list[DataCoverageMetric] = Field(default_factory=list)
    common_weaknesses: list[str] = Field(default_factory=list)
    summary: str


class AssignmentAnalysisResponse(BaseModel):
    report_id: str
    assignment_id: str
    assignment_title: str
    course_id: str
    course_name: str
    class_id: str
    class_name: str
    student_id: str
    student_name: str
    generated_at: str
    summary: str
    code_structure: CodeStructureSummary
    scores: list[AssignmentScore]
    findings: list[AssignmentFinding]
    evidence_snippets: list[CodeEvidenceSnippet] = Field(default_factory=list)
    capability_evidence: list[CapabilityEvidence]
    analysis_trace: list[AnalysisTraceStep] = Field(default_factory=list)
    improvement_tasks: list[str]
    citations: list[Citation] = Field(default_factory=list)
    access_scope: str = "demo"
    ai_generated: bool = True


class AssignmentDashboardResponse(BaseModel):
    assignment_id: str
    assignment_title: str
    course_id: str
    course_name: str
    class_id: str
    class_name: str
    generated_at: str
    submitted_count: int
    total_students: int
    average_score: int
    metrics: list[AssignmentDashboardMetric]
    dimension_averages: list[AssignmentScore]
    common_findings: list[AssignmentFinding]
    anomalies: list[AssignmentAnomaly] = Field(default_factory=list)
    teaching_suggestions: list[TeachingSuggestion]
    class_profile: ClassAbilityProfile
    reports: list[AssignmentReportSummary]
    access_scope: str = "demo"
    ai_generated: bool = True

File: frontend/src/shared/types/assignments.ts
export type AssignmentScore = {
  dimension: string;
  score: number;
  summary: string;
  evidence: string[];
};

export type AssignmentFinding = {
  severity: "high" | "medium" | "low" | string;
  title: string;
  detail: string;
  suggestion: string;
};

export type CodeEvidenceSnippet = {
  path: string;
  module: string;
  capability: string;
  line_start: number;
  line_end: number;
  snippet: string;
};

export type CapabilityEvidence = {
  dimension: string;
  evidence: string;
  source: string;
};

export type AnalysisTraceStep = {
  node: string;
  title: string;
  status: string;
  summary: string;
  evidence: string[];
};

export type CodeStructureSummary = {
  file_count: number;
  entry_files: string[];
  test_files: string[];
  documentation_files: string[];
  config_files: string[];
  detected_frameworks: string[];
  detected_capabilities: string[];
  risk_signals: string[];
};

export type Citation = {
  title: string;
  source_type: string;
  snippet: string;
};

export type AssignmentReport = {
  report_id: string;
  assignment_id: string;
  assignment_title: string;
  course_id: string;
  course_name: string;
  class_id: string;
  class_name: string;
  student_id: string;
  student_name: string;
  generated_at: string;
  summary: string;
  code_structure: CodeStructureSummary;
  scores: AssignmentScore[];
  findings: AssignmentFinding[];
  evidence_snippets: CodeEvidenceSnippet[];
  capability_evidence: CapabilityEvidence[];
  analysis_trace: AnalysisTraceStep[];
  improvement_tasks: string[];
  citations: Citation[];
  access_scope: string;
  ai_generated: boolean;
};

export type AssignmentReportSummary = {
  report_id: string;
  student_id: string;
  student_name: string;
  overall_score: number;
  status: string;
  summary: string;
};

export type AssignmentItem = {
  assignment_id: string;
  title: string;
  course_id: string;
  course_name: string;
  class_id: string;
  class_name: string;
  description: string;
  rubric_id?: string | null;
  created_at: string;
  submitted_count: number;
  access_scope: string;
};

export type AssignmentListResponse = {
  assignments: AssignmentItem[];
};

export type AssignmentExportResponse = {
  assignment_id: string;
  filename: string;
  content_type: string;
  markdown: string;
  generated_at: string;
  access_scope: string;
};

export type AssignmentDashboardMetric = {
  label: string;
  value: string;
  trend?: string | null;
};

export type TeachingSuggestion = {
  knowledge_point: string;
  class_evidence: string;
  suggested_activity: string;
  practice_task: string;
  expected_improvement: string;
};

export type AssignmentAnomaly = {
  severity: "high" | "medium" | "low" | string;
  title: string;
  affected_students: string[];
  evidence: string;
  suggested_action: string;
};

export type AbilityHeatmapCell = {
  student_id: string;
  student_name: string;
  dimension: string;
  score: number;
  level: "strong" | "stable" | "weak" | string;
};

export type DirectionDistributionItem = {
  direction: string;
  count: number;
  ratio: number;
};

export type DataCoverageMetric = {
  label: string;
  covered: number;
  total: number;
  ratio: number;
};

export type ClassAbilityProfile = {
  heatmap: AbilityHeatmapCell[];
  direction_distribution: DirectionDistributionItem[];
  data_coverage: DataCoverageMetric[];
  common_weaknesses: string[];
  summary: string;
};

export type AssignmentDashboard = {
  assignment_id: string;
  assignment_title: string;
  course_id: string;
  course_name: string;
  class_id: string;
  class_name: string;
  generated_at: string;
  submitted_count: number;
  total_students: number;
  average_score: number;
  metrics: AssignmentDashboardMetric[];
  dimension_averages: AssignmentScore[];
  common_findings: AssignmentFinding[];
  anomalies: AssignmentAnomaly[];
  teaching_suggestions: TeachingSuggestion[];
  class_profile: ClassAbilityProfile;
  reports: AssignmentReportSummary[];
  access_scope: string;
  ai_generated: boolean;
};

export type AssignmentUploadArchivePayload = {
  assignmentId?: string;
  assignmentTitle: string;
  archive: File;
  courseId?: string;
  classId?: string;
  studentId?: string;
  rubricId?: string;
  repositoryUrl?: string;
  description?: string;
};

export type RepositoryAnalysisPayload = {
  assignmentId?: string;
  assignmentTitle: string;
  courseId?: string;
  classId?: string;
  studentId?: string;
  repositoryUrl: string;
  description?: string;
};
